server/middleware.go

```
@@ -21,14 +26,49 @@ func jwtMiddleware(runtime *runtime.Runtime) gin.HandlerFunc {
 			jwt := authHeaders[1]
 			// use an env variable as jwt secret as upposed to using a stateful secret stored in
 			// database that is unique to every user and session
-			valid, userAddr, gErr := authJwtVerify(jwt, os.Getenv("JWT_SECRET"), runtime)
-			if gErr != nil {
-				c.JSON(http.StatusUnauthorized, gin.H{"error": gErr})
+			valid, userId, err := authJwtVerify(jwt, os.Getenv("JWT_SECRET"), runtime)
+			if err != nil {
+				c.JSON(http.StatusUnauthorized, gin.H{"error": err.Error()})
+				return
+			}
+
+			if !valid {
+				c.JSON(http.StatusUnauthorized, gin.H{"error": "invalid jwt"})
 				return
 			}
 
-			c.Set("authenticated", valid)
-			c.Set("user_addr", userAddr)
+			c.Set("user_id", userId)
+		} else {
+			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid authorization header format"})
+			return
+		}
+		c.Next()
+	}
+}
+
+func jwtOptional(runtime *runtime.Runtime) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		header := c.GetHeader("Authorization")
+		if header != "" {
```

if the jwt is optional, i wonder if we should even bother verifying the header. maybe we should fail silently instead.

example:
1. user requests a public page (not theirs) at gallery.so/cool_user
2. they haven't logged in in a while, so their JWT has expired
3. the request to fetch public gallery data for cool_user fails, and user is hit with error page

in theory, the user should have been able to see cool_user's profile, despite their auth status

---

it feels unusual to have middleware for something that is optional. if we deprecate `jwtOptional`, handlers should be cleaned up to only be used for the required case, which makes sense:

```js
	nftsGroup.GET("/get", getNftById(pRuntime))
	nftsGroup.GET("/user_get", getNftsForUser(pRuntime))
	nftsGroup.POST("/update", jwtRequired(pRuntime), updateNftById(pRuntime))
```

---

Currently the JWT optional middleware is a middleware for the endpoints that react differently to authenticated vs non-authenticated users. It will insert the auth status and user ID into the context so that future handlers can access them without blocking access to any future handlers.

We could have only one jwt middleware like we used to that just inserts the auth status into the context but this would require each handler to handle it's own authentication protection (e.g. the update collection handler will first have to grab the auth status from the context and check if a user is authed before continuing). I believe this is a bit out of the scope of most handlers. The handlers that react differently to different auth statuses, however, I believe should not have to verify jwt's themselves and should just have that status handed to them from the context. Let me know if this is incorrect or if there is a better solution you have in mind.

> in theory, the user should have been able to see cool_user's profile, despite their auth status

Currently, this middleware will not stop the user from reaching the next handler even if they are not authed/jwt invalid. Therefore, even a user with an expired JWT will continue through the `jwtOptional` middleware. Is this alright?

> if the jwt is optional, i wonder if we should even bother verifying the header. maybe we should fail silently instead.

Will do!

---

@bennycio this is perfect, thank you for laying it out so clearly!